mistralrs-core/src/utils/tokens.rs

```diff
@@ -14,44 +14,30 @@ enum TokenRetrievalError {
 
 /// This reads a token from a specified source. If the token cannot be read, a warning is logged with `tracing`
 /// and *no token is used*.
-pub(crate) fn get_token(source: &TokenSource) -> Result<String> {
-    Ok(match source {
-        TokenSource::Literal(data) => data.clone(),
-        TokenSource::EnvVar(envvar) => {
-            let tok = env::var(envvar);
-            if let Ok(tok) = tok {
-                tok
-            } else {
-                info!("Could not load token at {envvar:?}, using no HF token.");
-                "".to_string()
-            }
-        }
-        TokenSource::Path(path) => {
-            let tok = fs::read_to_string(path);
-            if let Ok(tok) = tok {
-                tok
-            } else {
-                info!("Could not load token at {path:?}, using no HF token.");
-                "".to_string()
-            }
-        }
+pub(crate) fn get_token(source: &TokenSource) -> Result<Option<String>> {
+    fn skip_token(input: &str) -> Option<String> {
+        info!("Could not load token at {input:?}, using no HF token.");
+        None
+    }
+
+    let token = match source {
```

Comment:
What about, if `EnvVar` or `Path` fail, we try `Cache`. So, the graph would look like this:
- Literal -> `return Some` <--- Same as current
- EnvVar | Path (maybe `return Some`) or -> notify, try CacheToken (maybe `return Some`) -> notify, None
- CacheToken (maybe `return Some`) -> notify, None  <--- Same as current

Reply:
I had a go at it, but the branches for `EnvVar` / `Path` calling `try_cache()` won't propagate the error if encountered, or at least not with the awkward looking approach I put together to share the cache logic as fallback:

```rs
pub(crate) fn get_token(source: &TokenSource) -> Result<Option<String>> {
    fn skip_token() -> Option<String> {
        info!("Proceeding without an HF token.");
        None
    }

    fn log_token_missing(location: &str) {
        info!("Could not find a token at '{location:?}'.");
    }

    fn try_cache() -> Result<Option<String>> {
        let home = format!(
            "{}/.cache/huggingface/token",
            dirs::home_dir()
                .ok_or(TokenRetrievalError::HomeDirectoryMissing)?
                .display()
        );

        Ok(fs::read_to_string(home.clone()).ok().or_else(|| {
            log_token_missing(&home);
            skip_token()
        }))
    }

    let token = match source {
        TokenSource::Literal(data) => Some(data.clone()),
        TokenSource::EnvVar(envvar) => env::var(envvar)
            .ok()
            .ok_or_else(|| {
                log_token_missing(envvar);
                try_cache()
            })
            .ok(),
        TokenSource::Path(path) => fs::read_to_string(path)
            .ok()
            .ok_or_else(|| {
                log_token_missing(path);
                try_cache()
            })
            .ok(),
        TokenSource::CacheToken => try_cache()?,
        TokenSource::None => None,
    };

    Ok(token.map(|s| s.trim().to_string()))
}
```

> What do you think?

I am not sure about the implicit fallback, I guess it's ok but if there is no actual user demand for it I would probably just keep it simple as-is to maintain?

Reply:
Yeah, implicit behavior is generally bad. I think it's fine the way it is. Thanks!